gnd/src/main/java/com/google/android/gnd/ui/home/featuresheet/RecordListViewModel.java

```diff
@@ -57,23 +71,48 @@ public void loadRecordSummaries(Feature feature, Form form) {
         feature.getProject(), feature.getFeatureType().getId(), form.getId(), feature.getId());
   }
 
+  /**
+   * Attempts to fetch a list of records based on the contents of a {@link RecordSummaryRequest}.
+   *
+   * @param request A record summary request. A triple of project, featureId, and formId.
+   * @return A list containing fetched records with forms that satisfy the formId
+   *     provided in the request.
+   */
+  private Single<List<Record>> fetchRecordSummaries(RecordSummaryRequest request) {
+    // TODO: Only fetch records with current formId.
+    return dataRepository
+        .getRecordSummaries(request.project.getId(), request.featureId)
+        .map(
+            records ->
+                stream(records)
+                    .filter(record -> record.getForm().getId().equals(request.formId))
+                    .collect(Collectors.toList()));
+  }
+
+  private void onRecordSummaryError(Throwable t) {
+    // TODO: Show an appropriate error message to the user.
+    Log.d(TAG, "Failed to fetch record summaries.", t);
+  }
+
   private void loadRecords(Project project, String featureTypeId, String formId, String featureId) {
     Optional<Form> form = project.getFeatureType(featureTypeId).flatMap(pt -> pt.getForm(formId));
     if (!form.isPresent()) {
       // TODO: Show error.
       return;
     }
     // TODO: Use project id instead of object.
-    // TODO(#24): Fix leaky subscriptions!
-    disposeOnClear(
-        dataRepository
-            .getRecordSummaries(project.getId(), featureId)
-            .subscribe(
-                // TODO: Only fetch records w/current formId.
-                records ->
-                    recordSummaries.setValue(
-                        stream(records)
-                            .filter(record -> record.getForm().getId().equals(formId))
-                            .collect(Collectors.toList()))));
+    recordSummaryRequests.onNext(new RecordSummaryRequest(project, featureId, formId));
+  }
+
+  class RecordSummaryRequest {
+    public Project project;
+    public String featureId;
+    public String formId;
```

Comment:
Please make these final so objects can be treated as immutable.

Reply:
Done!